### PR TITLE
[Feature](bangc-ops): add roialign_forward binary op

### PR DESCRIPTION
**Pull request by liuyuan1-v**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

roialign_forward二进制适配mluops

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [ ] diff1: diff1 <= 3e-3
- [ ] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block <br> U1 <br> U4       |
|        3       |Layouts                               |          NHWC 、NCHW、ARRAY etc      |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float etc           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/int8                   |          |          |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |          |
|Memory leak check    |Test result                       |          |          |
|Code coverage check  |Test result                       |          |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|op_name|    |    |     |    |    |    |     |
|op_name|    |    |     |    |    |    |     |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|op_name|    |    |    |    |    |    |     |
|op_name|    |    |    |    |    |    |     |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

**Review by cathyzhang222 — APPROVED** [review]

File: bangc-ops/mlu_op.h [hunk 1]
```diff
@@ -928,6 +928,15 @@ typedef struct mluOpReduceStruct *mluOpReduceDescriptor_t;
  */
 typedef struct mluOpTransposeStruct *mluOpTransposeDescriptor_t;
 
+/*! The descriptor of RoiAlignForward function that holds parameter information.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:931]: RoiAlignForward 是指的mluOpRoiAlignForward接口吧，这里写成  ::mluOpRoiAlignForward

File: bangc-ops/mlu_op.h [hunk 2]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
                        const mluOpTensorDescriptor_t bottom_grad_desc,
                        void *bottom_grad);
 
+// Group:RoiAlignForward
+/*!
+ * @brief Creates a descriptor pointed by \b desc for ::mluOpRoiAlignForward operation, and
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3838]: 去掉operation，这里是接口，不是算子

File: bangc-ops/mlu_op.h [hunk 3]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * allocates memory for holding the information about the operation. The information
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3839]: 这里operation改为function

File: bangc-ops/mlu_op.h [hunk 4]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * is defined in ::mluOpRoiAlignForwardDescriptor_t.
+ *
+ * @param[in] desc
+ * A host pointer to the descriptor that holds information about ::mluOpRoiAlignForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3843]: 去掉operation

File: bangc-ops/mlu_op.h [hunk 5]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - You need to call the ::mluOpDestroyRoiAlignForwardDescriptor function to destroy the descriptor.
+ *
+ * @par Note
+ * - None.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - None.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3860]: 参照https://wiki.cambricon.com/pages/viewpage.action?pageId=255658083补全其他API字段

File: bangc-ops/mlu_op.h [hunk 6]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ */
+
+mluOpStatus_t MLUOP_WIN_API
+mluOpCreateRoiAlignForwardDescriptor(mluOpRoiAlignForwardDescriptor_t *desc);
+
+// Group:RoiAlignForward
+/*!
+ * @brief Initializes the descriptor \b desc that is previously created with
+ * ::mluOpCreateRoiAlignForwardDescriptor function, and sets roialign information
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3869]: roialign 是否写成RoiAlign更合适？

File: bangc-ops/mlu_op.h [hunk 7]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * to the descriptor \b desc. The information includes the number of the roialign
+ * feature map height \b pooled_height, the roialign feature map width \b pooled_width,
+ * the sampling_ratio for each boxes \b sampling_ratio, the of the spatial_scale for
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3872]: the of the spatial_scale for  这里应该是写错了，少了内容，请确认一下

File: bangc-ops/mlu_op.h [hunk 8]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * The descriptor of the roialign operation. For detailed information,
+ * see ::mluOpRoiAlignForwardDescriptor_t.
+ * @param[in] pooled_height
+ * The height of output feature map.
+ * @param[in] pooled_width
+ * The width of output feature map.
+ * @param[in] sampling_ratio
+ * The number of sampling points in the grid used to compute the output.
+ * @param[in] spatial_scale
+ * The spatial scale of each regions of interest in the output.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3890]: regions of interest 》ROI

File: bangc-ops/mlu_op.h [hunk 9]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Note
+ * - None.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - None.
+ */
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3905]: 按照 https://wiki.cambricon.com/pages/viewpage.action?pageId=255658083 
中API的规范补全其他字段

File: bangc-ops/mlu_op.h [hunk 10]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+mluOpStatus_t MLUOP_WIN_API
+mluOpSetRoiAlignForwardDescriptor(mluOpRoiAlignForwardDescriptor_t desc,
+                                  const int pooled_height,
+                                  const int pooled_width,
+                                  const int sampling_ratio,
+                                  const float spatial_scale,
+                                  const bool aligned);
+// Group:RoiAlignForward
+/*!
+ * @brief Initializes the roialign descriptor \b roialign_desc that is previously created with
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3916]: 确定下roialign 修改为RoIAlign

File: bangc-ops/mlu_op.h [hunk 11]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * The descriptor of the roialign operation. For detailed information,
+ * see ::mluOpRoiAlignForwardDescriptor_t.
+ * @param[in] pooled_height
+ * The height of output feature map. The value of this parameter should be greater than 0.
+ * @param[in] pooled_width
+ * The width of output feature map. The value of this parameter should be greater than 0.
+ * @param[in] sampling_ratio
+ * The number of sampling points in the grid used to compute the output.
+ * @param[in] spatial_scale
+ * The spatial scale of each regions of interest in the output.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3934]: regions of interest > ROI

File: bangc-ops/mlu_op.h [hunk 12]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM
+ *
+ * @par Note
+ * - None.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - None.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3952]: 参照wiki补全其他字段，例如@Scale Limitation 等

File: bangc-ops/mlu_op.h [hunk 13]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+                                     const int pool_mode,
+                                     const bool aligned);
+
+// Group:RoiAlignForward
+/*!
+ * @brief Destroys a roialign descriptor \b desc that is previously created with the
+ * ::mluOpCreateRoiAlignForwardDescriptor function.
+ *
+ * The roialign descriptor is defined in ::mluOpRoiAlignForwardDescriptor_t
+ * and holds the information about the ::mluOpRoiAlignForward operation.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3969]: about ::mluOpRoiAlignForward.

File: bangc-ops/mlu_op.h [hunk 14]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @param[in] desc
+ * The roialign descriptor to be destroyed.
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_EXECUTION_FAILED
+ *
+ * @par Note
+ * - You need to call this function after calling the ::mluOpRoiAlignForward.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3978]: 去掉the

File: bangc-ops/mlu_op.h [hunk 15]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - None.
+ */
+
+mluOpStatus_t MLUOP_WIN_API
+mluOpDestroyRoiAlignForwardDescriptor(mluOpRoiAlignForwardDescriptor_t desc);
+
+// Group:RoiAlignForward
+/*!
+ * @brief Computes the output of images \b output_image based on the input \b input_tensor and
+ * bounding boxes \b boxes to perform ::mluOpRoiAlignForward operation. To use maximum pooling mode
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3995]: 去掉operation

File: bangc-ops/mlu_op.h [hunk 16]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * or average pooling mode in this operation, call ::mluOpRoiAlignForward_v2.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:3996]: 去掉 in this operation

File: bangc-ops/mlu_op.h [hunk 17]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Deprecated
+ * - ::mluOpRoiAlignForward is deprecated and will be removed in the future release. It is
+ * recommended to use ::mluOpRoiAlignForward_v2 instead, which supports both maximum and average modes.
+ *
+ * @param[in] handle
+ * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4003]: Handle to an MLUOP context

File: bangc-ops/mlu_op.h [hunk 18]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * ::mluOpRoiAlignForward operation. For detailed information, see ::mluOpHandle_t.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4004]: 去掉operation

File: bangc-ops/mlu_op.h [hunk 19]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * @param[in] roialign_desc
+ * The descriptor of the roialign operation. For detailed information, see
+ * ::mluOpRoiAlignForwardDescriptor_t.
+ * @param[in] input_desc
+ * The descriptor of the input tensor in the roialign process. For detailed
+ * information, see ::mluOpTensorDescriptor_t.
+ * @param[in] input
+ * Pointer to the MLU memory that stores the input tensor.
+ * @param[in] boxes_desc
+ * Descriptor of boxes, containing dimension and the layout of boxes.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4014]: The descriptor of boxes, which contains dimension and the layout of boxes.

File: bangc-ops/mlu_op.h [hunk 20]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] boxes
+ * Pointer to the MLU memory that stores the boxes tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4017]: that stores \b boxes tensor.

File: bangc-ops/mlu_op.h [hunk 21]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in
+ * ::mluOpRoiAlignForward operation. For detailed information, see ::mluOpHandle_t.
+ * @param[in] roialign_desc
+ * The descriptor of the roialign operation. For detailed information, see
+ * ::mluOpRoiAlignForwardDescriptor_t.
+ * @param[in] input_desc
+ * The descriptor of the input tensor in the roialign process. For detailed
+ * information, see ::mluOpTensorDescriptor_t.
+ * @param[in] input
+ * Pointer to the MLU memory that stores the input tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4012]: stores \b input tensor.

File: bangc-ops/mlu_op.h [hunk 22]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * @param[in] boxes_desc
+ * Descriptor of boxes, containing dimension and the layout of boxes.
+ * For detailed information, see ::mluOpTensorDescriptor_t.
+ * @param[in] boxes
+ * Pointer to the MLU memory that stores the boxes tensor.
+ * @param[in] output_desc
+ * Descriptor of output, containing dimension and the layout of output.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4019]: The descriptor of output, which contains dimension and the layout of output.

File: bangc-ops/mlu_op.h [hunk 23]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * @param[out] output
+ * Pointer to the MLU memory that stores the output tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4021]: that stores \b output tensor.

File: bangc-ops/mlu_op.h [hunk 24]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Return
+ * - ::MLUOP_STATUS_SUCCESS, ::MLUOP_STATUS_BAD_PARAM.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4024]: 去掉句尾句号

File: bangc-ops/mlu_op.h [hunk 25]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - This function supports the following data types for input tensor \b input, boxes tensor
+ *   \b boxes, and output tensor \b output tensor. Data type of all tensors should be the same.
+ *   - input tensor: half, float.
+ *   - boxes tensor: half, float.
+ *   - output tensor: half, float.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4031]: 这三行去掉句尾句号

File: bangc-ops/mlu_op.h [hunk 26]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layout of \b input, \b boxes, \b output are as follows:
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4034]: data layouts

File: bangc-ops/mlu_op.h [hunk 27]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *   - input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY, only supports 2D tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4036]: \p MLUOP_LAYOUT_ARRAY, which only supports 2D tensor.

File: bangc-ops/mlu_op.h [hunk 28]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *   - output tensor: \p MLUOP_LAYOUT_NHWC.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4037]: 这三行去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 29]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Scale Limitation
+ * - The input tensor and output tensor must have four dimensions.
+ * - data type of half is not recommended due to low precision.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4041]: The data type half

File: bangc-ops/mlu_op.h [hunk 30]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - Size of the lowest dimension of input tensor and output tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4042]: The size

File: bangc-ops/mlu_op.h [hunk 31]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - The boxes tensor must have two dimensions.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4043]: \b boxes tensor

File: bangc-ops/mlu_op.h [hunk 32]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - Size of the highest dimension of output tensor and boxes tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4044]: The size

File: bangc-ops/mlu_op.h [hunk 33]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - None.
+ *
+ * @par Note
+ * - None.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - The example of the ::mluOpRoiAlignForward operation is as follows:
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4060]: 去掉operation

File: bangc-ops/mlu_op.h [hunk 34]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+                     const void *input,
+                     const mluOpTensorDescriptor_t boxes_desc,
+                     const void *boxes,
+                     const mluOpTensorDescriptor_t output_desc,
+                     void *output);
+
+// Group:RoiAlignForward
+/*!
+ * @brief Computes the output feature map \b output based on the input feature map \b input and
+ * bounding boxes \b boxes to perform this operation. Compared with ::mluOpRoiAlignForward, in addition to
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4091]: operation > function

File: bangc-ops/mlu_op.h [hunk 35]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * supporting average pooling mode, ::mluOpRoiAlignForward_v2 also supports maximum pooling mode with two
+ * more output \b argmax_x and \b argmax_y.
+ * @param[in] handle
+ * Handle to a Cambricon MLUOP context that is used to manage MLU devices and queues in
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4095]: Handle to an MLUOP context

File: bangc-ops/mlu_op.h [hunk 36]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * ::mluOpRoiAlignForward_v2 operation. For detailed information, see ::mluOpHandle_t.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4096]: 去掉operation

File: bangc-ops/mlu_op.h [hunk 37]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * @param[in] roialign_desc
+ * The descriptor of the roialign operation. For detailed information, see
+ * ::mluOpRoiAlignForwardDescriptor_t.
+ * @param[in] input_desc
+ * The descriptor of the input tensor in the roialign process. For detailed
+ * information, see ::mluOpTensorDescriptor_t.
+ * @param[in] input
+ * Pointer to the MLU memory that stores the input tensor.
+ * @param[in] boxes_desc
+ * The descriptor of the region proposals tensor. For detailed information, see
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4106]: 这里是写错了嘛？是The descriptor of the \b boxes tensor.

File: bangc-ops/mlu_op.h [hunk 38]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * ::mluOpTensorDescriptor_t.
+ * @param[in] boxes
+ * Pointer to the MLU memory that stores the region proposals tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4109]: 请确认一下，这里是写错了嘛？that stores \b boxes tensor.

File: bangc-ops/mlu_op.h [hunk 39]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Data Type
+ * - This function supports the following data types for input tensor \b input, boxes tensor
+ *   \b boxes, output tensor \b output tensor, argmax_x tensor \b argmax_x and argmax_y tensor
+ *   \b argmax_y. Data type of all tensors should be the same.
+ *   - input tensor: half, float.
+ *   - boxes tensor: half, float.
+ *   - output tensor: half, float.
+ *   - argmax_x tensor: half, float.
+ *   - argmax_y tensor: half, float.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4141]: 去掉每句后面的句号

File: bangc-ops/mlu_op.h [hunk 40]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Data Layout
+ * - The supported data layout of \b input, \b boxes, \b output, \b argmax_x and \b argmax_y are as
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4144]: data layouts

File: bangc-ops/mlu_op.h [hunk 41]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *   follows:
+ *   - input tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - boxes tensor: \p MLUOP_LAYOUT_ARRAY, only supports 2D tensor.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4147]: \p MLUOP_LAYOUT_ARRAY, which only supports 2D tensor.

File: bangc-ops/mlu_op.h [hunk 42]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *   - output tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - argmax_x tensor: \p MLUOP_LAYOUT_NHWC.
+ *   - argmax_y tensor: \p MLUOP_LAYOUT_NHWC.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4150]: 去掉句尾的句号

File: bangc-ops/mlu_op.h [hunk 43]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ *
+ * @par Scale Limitation
+ * - The \b input tensor, \b output tensor, \b argmax_x tensor and \b argmax_y tensor must have four dimensions.
+ * - \b input data type of half is not recommended due to low precision.
+ * - Size of the lowest dimension of \b input tensor and \b output tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4155]: The size

File: bangc-ops/mlu_op.h [hunk 44]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - Size of the lowest dimension of \b input tensor and \b argmax_x tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4156]: The size

File: bangc-ops/mlu_op.h [hunk 45]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - Size of the lowest dimension of \b input tensor and \b argmax_y tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4157]: The size

File: bangc-ops/mlu_op.h [hunk 46]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * - The \b boxes tensor must have two dimensions.
+ * - Size of the highest dimension of \b output tensor and \b boxes tensor must be the same.
+ * - Size of the highest dimension of \b argmax_x tensor and \b boxes tensor must be the same.
+ * - Size of the highest dimension of \b argmax_y tensor and \b boxes tensor must be the same.
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4161]: 这三个size前都加上The size

File: bangc-ops/mlu_op.h [hunk 47]
```diff
@@ -3824,6 +3833,387 @@ mluOpPsRoiPoolBackward(mluOpHandle_t handle,
...
+ * @par Note
+ * - When \b input contains NaN:
+ *   If \b pool_mode is maximum pooling_mode, \b output is positive saturation value on MLU200 series
+ *   and \b output gets more NaN than ieee754 on MLU300 series and CE3226.
+ *
+ * @par Requirements
+ * - None.
+ *
+ * @par Example
+ * - The example of the ::mluOpRoiAlignForward operation is as follows:
```
Comment (GraceGuanH) [inline bangc-ops/mlu_op.h:4181]: 去掉operation

**Review by GraceGuanH — COMMENTED** [review]
Please check the comments.

**Review by GraceGuanH — APPROVED** [review]

**Review by LingyanZhang-HAH — APPROVED** [review]

**Pull request closed without merging**